### ISSUE TITLE
Auto resolve after `n` days of inactivity

### ISSUE THREAD

**Issue by pranavrajs** [chatwoot/chatwoot#418]
**Is your enhancement request related to a problem? Please describe.**
Add a feature to resolve the conversation after `n` days of inactivity. If I'm waiting for a response from the customer and he has not responded in `n` days, the conversation needs to resolved automatically.

**Describe the solution you'd like**
Add a setting in account `auto_resolve_duration` which can be set by an admin, if there is no message or any activity happening in that duration, mark the conversation as resolved.

### Acceptance criteria

- Add a DB field on account `auto_resolve_duration`
- The conversation should auto-resolve if there is no activity.

**Comment by pranavrajs** [comment]
@sojan-official De-prioritizing this and moving this to next sprint.

**Comment by akashdotsrivastava** [comment]
@pranavrajs 
Hi. I am new to chatwoot.
If this is still needed, can I take this issue?

**Comment by pranavrajs** [comment]
@akashdotsrivastava Yes, it is still needed. You can take the issue. Let me know if you need anything from me.

**Comment by akashdotsrivastava** [comment]
@pranavrajs 
While working on this, I realized the time metric we should use to auto-resolve conversations, while not there in `develop` yet, has been introduced in #1281 (`last_activity_at` in `conversations`). I can do my changes for this issue on top of those changes if that's okay.

But I saw a couple of test cases failing for #1281, so I added a [PR fixing those test cases](https://github.com/chatwoot/chatwoot/pull/1302), targeted at the original branch (`fix/1207-timestamp-ordering`).

**Comment by pranavrajs** [comment]
@akashdotsrivastava This sounds like a good idea, please feel free to build on top of #1281. I will try to get it merged today. Thanks for the spec fixes.

**Comment by akashdotsrivastava** [comment]
@pranavrajs Thanks

Couple of questions.
 
1. I understand the description says adding the auto_resolve_duration field to accounts. But from the looks of it, I don't really see many actual settings like properties for accounts, on the UI. I see most of them at the inbox level. So should we still add this setting to an account, or to an inbox?

2. What exactly is going to trigger the auto-resolve? 

    Usually, in systems that do this kind of auto status change, some kind of a periodic task (Cron job) takes care of this. We can add a rake task triggering a job that updates the status, adding that rake task to the cron, at any kind of interval. Again, because I am new to chatwoot, do we do crons in chatwoot? Or should I just add the task and be done for now?

   The second option, we can trigger a background job to perform_later(after n days) when a conversation is created. In the job itself, we can check that if there has been more activity, and then reset it to perform later after the rest amount of time, and so on, and resolve and finish job otherwise. But in that case it depends on sidekiq -> redis, which is not very reliable for setting up a job in queue for days (usually 20-30), in case redis data is lost.

   So basically the question is what extra worker is going to take care of this auto resolve?

**Comment by sojan-official** [comment]
@akashdotsrivastava 

1. We can keep it `account settings`, for now, The inbox settings are starting to getting cluttered. We will move this to workflows when we introduce that feature later. 

2.  At the moment we don't have any sort of cron jobs in chatwoot. Sidekiq handles repetitive tasks.  
I get the concern you mentioned with the dependency on Redis.  But let's go with the approach you mentioned, leveraging the `perform_later` for now.  Since we have chatwoot installations running over PaaS like Heroku, we would need to investigate and document to introduce a better strategy for scheduled jobs.  We can refactor the code at that point.

**Comment by akashdotsrivastava** [comment]
@sojan-official Got it. Thanks.
